batch-l2.md

```diff
@@ -0,0 +1,1494 @@
+In particular, the environment variable that tells the containers which one is the main node can be configured on your `MultiNodeJobDefinition` as follows:
+
+```ts
+const multiNodeJob = new batch.MultiNodeJobDefinition(this, 'JobDefinition', {
+  mainNode: 0,
+  instanceType: ec2.InstanceType.of(ec2.InstanceClass.A1, ec2.InstanceSize.LARGE),
+  containers: [
+    new MultiNodeContainer(this, 'mainMPIContainer', {
+      startNode: 0,
+      endNode: 5,
+      image: ecs.ContainerImage.fromRegistry('yourregsitry.com/yourMPIImage:latest'),
+      memoryLimitMiB: 2048,
+    }),
+  ],
+});
+
+// convenience method
+multiNodeJob.addContainer(this, 'secondContanerType', {
+  startNode: 6,
+  endNode: 10,
+  image: ecs.ContainerImage.fromRegistry('public.ecr.aws/amazonlinux/amazonlinux:latest'),
+  memoryLimitMiB: 2048,
+});
+```
+
+For an in-depth example on using MPI to perform numerical computations on Batch, see https://aws.amazon.com/blogs/compute/building-a-tightly-coupled-molecular-dynamics-workflow-with-multi-node-parallel-jobs-in-aws-batch/
+
+### Pass Parameters to a Job
+
+Batch allows you define parameters in your `JobDefinition`, which can be referenced in the container command. For example:
+
+```
+new batch.EcsJobDefinition(this, 'JobDefn', {
+  parameters: { echoParam: 'foobar' },
+  containerDefinition: new batch.ContainerDefinition(this, 'containerDefn', {
+    image: ecs.ContainerImage.fromRegistry('public.ecr.aws/amazonlinux/amazonlinux:latest'),
+    memoryLimitMiB: 2048,
+    compatibility: Compatibility.EC2,
+    command: [
+      'echo',
+      'Ref::echoParam',
+    ],
+  }),
+});
+```
+
+## API
+
+* `ManagedEc2ComputeEnvironment`
+    * Requires an L2 for `CfnPlacementGroup` in aws-ec2
+        * Or, we could use a `placementGroupArn` instead, but that’s sub-par
+* `UnmanagedEc2ComputeEnvironment`
+* `FargateComputeEnvironment`
+* `MultinodeJobDefinition`
+* `ContainerJobDefinition`
+* `JobQueue`
+* `SchedulingPolicy`
+
+```ts
+new ManagedEc2ComputeEnvironment(this, 'myEc2Env', {
+  images: [BatchMachineImage.of(ec2MachineImage, BatchImageType.ECS)], // required
+  name: 'specifiedName', // optional, defaults to CFN generated name
+  replaceComputeEnvironment: false, //optional, defaults to false
+  spot: true, // optional, defaults to false
+  enabled: true, // optional, defaults to true
+  updateTimeout: cdk.Duration.minutes(30), // optional, defaults to 30 (CFN default)
+  terimnateOnUpdate: false, // optional, defaults to false (CFN default)
+  allocationStrategy: batch.AllocationStrategy.SPOT_CAPACITY_OPTIMIZED, // optional (not spot exclusive)
+  spotBidPercentage: 20, // optional (Spot exclusive)
+  desiredvCpus: 3, // optional: default???
+  instanceTypes: [ec2.InstanceType.of(ec2.InstanceClass.M5AD, ec2.InstanceSize.LARGE)], // optional, defaults to ['Instance.OPTIMAL']
+  instanceRole: myRole, // optional, creates it if not supplied
+  serviceRole: myServiceRole, // optional, creates it if not supplied
```

Comment:
This comment is for all occurrences of `serviceRole` in this RFC.

For ALL managed CEs (ECS, EKS, Fargate) in Batch, if serviceRole is null at CreateCE API call time, Batch will create a Service-Linked-Role (SLR) on the fly for the customer or it will re use an existing one. So CDK should totally get out of the business of creating role/policy for serviceRole with Managed CEs and let the service use default SLR. CDK can keep that create role logic for Unmanaged type.

SLR is the more secure and less error prone method of giving Batch access to orchestrate jobs and scaling in customer's account.  It should be used by default for all managed CEs.